src/service/gateways/bitstamp.ts

```diff
@@ -0,0 +1,360 @@
+	
+	ConnectChanged = new Utils.Evt<Models.ConnectivityStatus>();
+    
+    private onMessage = (msg : Models.Timestamped<OrderBook | Ticker>) => {
+        if (msg.data.hasOwnProperty("id")) {
+            this.onTicker(<Models.Timestamped<Ticker>>msg);
+        }
+        else {
+            this.onOrderBookMessage(<Models.Timestamped<OrderBook>>msg);
+        }
+    };
+	
+	private static ConvertToMarketSide = (input : [string, string]) => 
+        new Models.MarketSide(parseFloat(input[0]), parseFloat(input[1]));
+        
+    private static ConvertToMarketSideList = (input : [string, string][]) => 
+        _(input).slice(0, 5).map(BitstampMarketDataGateway.ConvertToMarketSide).value();
+        
+	private onOrderBookMessage = (message : Models.Timestamped<OrderBook>) => {
+		var bids = BitstampMarketDataGateway.ConvertToMarketSideList(message.data.bids);
+        var asks = BitstampMarketDataGateway.ConvertToMarketSideList(message.data.asks);
+		this.MarketData.trigger(new Models.Market(bids, asks, message.time));
+	};
+    
+    private onTicker = (message : Models.Timestamped<Ticker>) => {
+        var trd = new Models.GatewayMarketTrade(message.data.price, message.data.amount,  message.time, false, null);
+        this.MarketTrade.trigger(trd);
+    };
+	
+     _log : Utils.Logger = Utils.log("tribeca:gateway:BitstampMD");
+    constructor(restClient: BitstampAuthenticatedClient, url: string) {
+		this._client = new PusherClient(url, "BitstampPusherClient", [["order_book", "data"], ["live_trades", "trade"]]);
+		this._client.ConnectChanged.on(c => this.ConnectChanged.trigger(c));
+		this._client.Message.on(this.onMessage);
+        
+        restClient.getFromEndpoint<OrderBook>("order_book/").then(b => 
+            this.onOrderBookMessage(new Models.Timestamped(b, moment.utc())))
+                .done();
+            
+        restClient.getFromEndpoint<MarketTransaction[]>("transactions/").then(ts => {
+            ts.forEach(t => {
+                var side = t.type === 0 ? Models.Side.Bid : Models.Side.Ask; // is this really the make side?
+                var trd = new Models.GatewayMarketTrade(parseFloat(t.price), parseFloat(t.amount), moment(t.date, "X"), true, side);
+                this.MarketTrade.trigger(trd);
+            });
+        }).done();
+    }
+}
+
+class BitstampAuthenticatedClient {
+    private static UserAgent = 'Mozilla/4.0 (compatible; Bitstamp node.js client)';
+    
+    constructor(
+        private baseUrl: string, 
+        private _customerId: string, 
+        private _apiKey: string, 
+        private _secret: string) { }
+        
+    private _lastTimeMs : number = 0;
+    private getNonce = () => {
+        var t = new Date().getTime();
+        if (t === this._lastTimeMs) {
+            this._lastTimeMs++;
+        }
+        else {
+            this._lastTimeMs = t*100;
+        }
+        
+        return this._lastTimeMs;
+    };
+    
+    private addAuthentication = (req: {}) => {
```

Comment:
Neither this nor the post method are not tested. I used https://github.com/askmike/bitstamp as a reference for the crypto functions.